modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1477,3 +1477,54 @@ def rebalance_partitions(cls, partitions):
             lengths = [part.length() for part in new_partitions[:, 0]]
             return new_partitions, lengths
         return partitions, None
+
+    @classmethod
+    def shuffle_partitions(
+        cls, partitions, sample_func, pivot_func, split_func, final_shuffle_func
+    ):
+        """
+        Return shuffled partitions.
+
+        Parameters
+        ----------
+        partitions : np.ndarray
+            The 2-d array of partitions to shuffle.
+        sample_func : Callable(pandas.DataFrame) -> Any
+            Function to sample partitions. Output of this function will be passed to ``pivot_func``
+            to determine pivots.
+        pivot_func : Callable(List[Any]) -> Any
+            Function to determine pivots from partitions. The pivots determined by this function
+            will be passed to ``split_func`` to split each partition.
+        split_func : Callable(pandas.DataFrame, Any) -> List[pandas.Dataframe]
+            Function that splits a partition based off of the pivots provided. Should return an
+            unpacked list of pandas Dataframes.
+        final_shuffle_func : Callable(pandas.DataFrame) -> pandas.DataFrame
+            Function that shuffles the data within each new partition.
+
+        Returns
+        -------
+        np.ndarray
+            A list of row-partitions that have been shuffled.
+        """
+        # Sample each partition
+        samples = cls.map_axis_partitions(1, partitions, sample_func, lengths=[None])
+        # Get each sample to pass in to the pivot function
+        samples = [sample.get() for sample in samples.flatten()]
+        pivots = pivot_func(samples)
+        # Convert our list of block partitions to row partitions
+        row_partitions = [
+            partition.force_materialization().list_of_block_partitions[0]
+            for partition in cls.row_partitions(partitions)
+        ]
+        # Gather together all of the sub-partitions
+        split_row_partitions = np.array(
+            [
+                partition.split(split_func, len(pivots) + 1, pivots)
+                for partition in row_partitions
+            ]
+        ).T
+        new_partitions = [
+            [cls._partition_class.put_splits(final_shuffle_func, splits)]
+            for splits in split_row_partitions
+        ]
```

Comment:
Could we have `split` return proper block partitions instead of just futures, and do a full-axis apply for the final shuffle? That way we could avoid having the `put_splits` function, which seems very similar to the existing full-axis `apply`.